os/skia/skia_surface.cpp

```diff
@@ -76,6 +82,40 @@ void SkiaSurface::create(int width, int height, const os::ColorSpaceRef& cs)
   swapBitmap(bmp);
 }
 
+void SkiaSurface::create(int width, int height,
+                         const os::SurfaceFormatData& sf,
+                         const unsigned char* data)
+{
+  destroy();
+
+  ASSERT(!m_surface)
+    ASSERT(width > 0);
+  ASSERT(height > 0);
+
+  SkColorType ct = SkiaSurface::deductSkColorType(sf);
+  SkAlphaType at = SkiaSurface::asSkAlphaType(sf.pixelAlpha);
+
+  SkBitmap bmp;
+  if (!bmp.tryAllocPixels(
+      SkImageInfo::Make(width, height, ct, at)))
+    throw base::Exception("Cannot create Skia surface");
+
+  if (data) {
+    // Convert 24bpp data to 32bpp data because Skia doesn't work with 24bpp
+    // images.
+    if (sf.bitsPerPixel == 24) {
+      for (int i=0,j=0; j<width * height * 3; i+=4,j+=3)
+        std::memcpy(&static_cast<unsigned char*>(bmp.getPixels())[i], &data[j], 3);
+    }
+    else
+      bmp.writePixels(SkPixmap(bmp.info(), data, size_t(width) * size_t(sf.bitsPerPixel / 8)));
```

Comment:
I'm not sure if it's clear enough what the `sf` parameter means, is the format of the resulting SkiaSurface, or the format of the input `data` buffer? Probably we need an extra parameter for the input `data` format.

Reply:
It is actually for both. The data must be arranged/formatted according to the SurfaceFormatData passed. Then if Surface::getFormat(SurfaceFormatData* formatData) is called on the created surface, it should fill the formatData with the original values (except the `format` member which is not used to create the surface, but anyway there is only one value...maybe this member is the confusing part?).
Hence, I don't think that adding an extra parameter will make things clear. Maybe just documenting the method would be enough.

Reply:
This is the confusing part: `sf.bitsPerPixel == 24`
This is the input data format, but not the output (the output will be 32 bpp, although not sure if kRGBA_8888_SkColorType or kBGRA_8888_SkColorType). I suppose here you are expecting the user to use the same shift/mask of RGB channels in 24bpp and 32bpp (?)

Anyway I think that if we expose the whole range of Skia formats in our laf-os API, we could completely remove the SurfaceFormatData. And in that case I'm not sure how the 24bpp would be specified for the input data.

Reply:
Other possibility to simplify the API right now (and avoid adding more complexity to the current PR), is to just not expose the pixel format and use a SkImageInfo::MakeN32() regular SkiaSurface with the image that is given from the clipboard, and do the conversion in `DragDataProviderWin::getImage()` to the RGBA `os::Surface`.

Reply:
I think what we are looking for is a function to convert the clip image (which can be obtained in several formats) into a standard os::SkiaSurface (N32), without exposing more color formats at the moment (because this is a more advanced feature for the laf-os library which is not required).

> Why would we do the conversion in `DragDataProviderWin`?

If you are getting a 24bpp image from macOS and Windows (which probably we do because clip tries to keep the original format) you can create a function that is called from both sides. We need a `convert_clip_image_to_surface()` (internal function/platform independent) that can be used in these getImages() implementations.

Reply:
What if, now that laf depends on clip, we replace this:
```c++
    virtual Ref<Surface> makeSurface(int width, int height, const os::SurfaceFormatData& sf, const unsigned char* data = nullptr) = 0;
```
with this
```c++
    virtual Ref<Surface> makeSurface(const clip::image_spec& spec, const clip::image& image) = 0;
```
And do the convert clip image to surface there?

Reply:
Yes, we can create that function. Anyway the spec is already inside the image (`image.spec()`):

    virtual Ref<Surface> makeSurface(const clip::image& image) = 0;

A couple of things about this:

1) Forward declare `clip::image` (don't include the clip header file from `os/system.h`)
2) After merging this PR I'll add a switch to avoid using the clip library from laf (and disabling the whole drag-and-drop stuff). But this is not part of this PR now.

Reply:
Yes, you can replace:
```
 Ref<Surface> makeSurface(int width, int height,
                          const os::SurfaceFormatData& sf,
                          const unsigned char* data);
```
with the new one that receives the `clip` image.

I'm not sure, but probably this makeSurface() can be defined in CommonSystem, because it doesn't depend on Skia (I guess).

Reply:
> I'm not sure, but probably this makeSurface() can be defined in CommonSystem, because it doesn't depend on Skia (I guess).

Actually the current makeSurface() doesn't depend on Skia either, isn't it? I would leave the method in System, but if you prefer to have it in CommonSystem, I'm fine. Just confirm which alternative should I go for.

Reply:
I'm not sure what do you mean with "current makeSurface"? The makeSurface() version created in this PR (that received data and format) must be gone and replaced with the clip one.

Reply:
Got it now. I was confused about the CommonSystem class, thank you for the clarification.

Reply:
Seems that I cannot call the "generic" makeSurface() because there isn't one. The only implementation of that makeSurface is in the SkiaSystem class, then when I try to call it from CommonSystem it is actually a pure virtual function and then I cannot call it (at least as far as I understood...).

Reply:
I've been trying to implement this in CommonSystem:
```c++
#if CLIP_ENABLE_IMAGE
  SurfaceRef makeSurface(const clip::image& image) override {
    return makeSurface(image.spec().width, image.spec().height);
  }
#endif
```
I'm just calling the "generic" makeSurface (if that is the one)...but I'm getting this compilation error:
```
C:\...\aseprite\laf\os\common\system.h(129): error C2660: 'os::CommonSystem::makeSurface': function does not take 2 arguments
  C:\...\aseprite\laf\os/common/system.h(128): note: see declaration of 'os::CommonSystem::makeSurface'
  C:\...\aseprite\laf\os/common/system.h(129): note: while trying to match the argument list '(const unsigned long, const unsigned long)'
```

Any idea?

Reply:
I've found out that the issue is because the generic makeSurface:
```c++
virtual Ref<Surface> System::makeSurface(int, int, const os::ColorSpaceRef& = nullptr) = 0;
```
Is declared abstract in `System`.

Then I guess that CommonSystem doesn't have any implementation to link to. 

So there are two options I think:
1) Implement the generic makeSurface in CommonSystem as a dummy method:
```c++
Ref<Surface> System::makeSurface(int, int, const os::ColorSpaceRef& = nullptr) { return nullptr; }
```
 And then implement the new makeSurface in CommonSystem as follows:
```c++
#if CLIP_ENABLE_IMAGE
  SurfaceRef makeSurface(const clip::image& image) override {
    return makeSurface(image.spec().width, image.spec().height);
  }
#endif
```

2) Implement the new makeSurface directly on SkiaSystem.

Reply:
I'm not in the computer now, please try doing `this->makeSurface(...)`

Reply:
> I'm not in the computer now, please try doing `this->makeSurface(...)`

Double check this, it doesn't work. Same compilation error as in https://github.com/aseprite/laf/pull/83#discussion_r1569245021